lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,96 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+void generateMemberwiseInit(ClassDecl *classDecl,
+                            SourceEditConsumer &EditConsumer,
+                            SourceManager &SM) {
+  
+  swift::SourceLoc bracesStart = classDecl->getBraces().Start;
+  
+  if (!bracesStart.isValid()) return;
+  
+  swift::SourceLoc targetLocation = bracesStart.getAdvancedLoc(1);
+  
+  if (!targetLocation.isValid()) return;
+  
+  SmallVector<std::string, 5> memberNameVector;
+  SmallVector<std::string, 5> memberTypeVector;
+  
+  for (auto varDecl : classDecl->getStoredProperties()) {
+    auto parentPatternBinding = varDecl->getParentPatternBinding();
+    if (!parentPatternBinding) continue;
+    
+    auto varDeclIndex =
+      parentPatternBinding->getPatternEntryIndexForVarDecl(varDecl);
+    
+    if (auto init = varDecl->getParentPatternBinding()->getInit(varDeclIndex)) {
+      if (init->getStartLoc().isValid())
+        continue;
+    }
+
+    StringRef memberName = varDecl->getName().str();
+    memberNameVector.push_back(memberName.str());
+    
+    std::string memberType = varDecl->getType().getString();
+    memberTypeVector.push_back(memberType);
+  }
```

Comment:
Can we move this for-loop into `canGenerateMemberwiseInit`?  Instead of returning `ClassDecl*`, `canGenerateMemberwiseInit` should populate `memberTypeVector` and `memberNameVector`. `isApplicable` returns true only when these two vectors aren't empty.  This way ensures that whenever the refactoring is available, we are going to change the users' codebase after applying it.

Reply:
I'm not clear about this. If I move the said for loop into `canGenerateMemberwiseInit`, `memberTypeVector` and `memberNameVector` will be local to that function. When `performChange` is called, it checks if `canGeneratememberWiseInit` returns a valid object and then calls `generateMemberwiseInit` which needs these vectors to emit the generated initializer. So my question is should we pass in references to the two vectors so that they are available out side the scope of the function and is it okay to have these vectors be populated twice?

Reply:
@mohitathwani yeah, you're right! We should pass `memberTypeVector` and `memberTypeVector ` as reference to `canGenerateMemberwiseInit` (maybe we should rename it to `colllectMemberForInit`. `isApplicable` will check whether these vectors are empty and `performChange` will use these two vectors for actual code change.